SpeechSandbox/Assets/Scripts/SpeechSandboxStreaming.cs

```diff
@@ -34,52 +34,140 @@ public class SpeechSandboxStreaming : MonoBehaviour
+
+    [Header("IAM Authentication")]
+    [Tooltip("The IAM apikey.")]
+    [SerializeField]
+    private string asst_iamApikey;
+    [Tooltip("The IAM url used to authenticate the apikey (optional). This defaults to \"https://iam.bluemix.net/identity/token\".")]
+    [SerializeField]
+    private string asst_iamUrl;
 
-    private string convo_username = "";
-    private string convo_password = "";
-    // Change convo_url if different from below
-    // Use this for TLS1.0 required by Unity as of 4/25/18
-    private string convo_url = "https://gateway-tls10.watsonplatform.net/conversation/api";
-    // Change  _conversationVersionDate if different from below
-    private string _conversationVersionDate = "2017-05-26";
-    private string convo_workspaceId = "";
+    #endregion
 
-    public Text ResultsField;
 
     private int _recordingRoutine = 0;
     private string _microphoneID = null;
     private AudioClip _recording = null;
     private int _recordingBufferSize = 1;
     private int _recordingHZ = 22050;
 
+    private SpeechToText _speechToText;
+    private Conversation _conversation;
 
-    void Start()
-    {
-        LogSystem.InstallDefaultReactors();
+    private IEnumerator createServices(){
 
+        Credentials stt_credentials = null;
         //  Create credential and instantiate service
-        Credentials stt_credentials = new Credentials(stt_username, stt_password, stt_url);
-        Credentials convo_credentials = new Credentials(convo_username, convo_password, convo_url);
+        if (!string.IsNullOrEmpty(stt_username) && !string.IsNullOrEmpty(stt_password))
+        {
+            //  Authenticate using username and password
+            stt_credentials = new Credentials(stt_username, stt_password, stt_serviceUrl);
+        }
+        else if (!string.IsNullOrEmpty(stt_iamApikey))
+        {
+            //  Authenticate using iamApikey
+            TokenOptions tokenOptions = new TokenOptions()
+            {
+                IamApiKey = stt_iamApikey,
+                IamUrl = stt_iamUrl
+            };
+
+            stt_credentials = new Credentials(tokenOptions, stt_serviceUrl);
+
+            while (!stt_credentials.HasIamTokenData())
+                yield return null;
+        }
+        else
+        {
+            throw new WatsonException("Please provide either username and password or IAM apikey to authenticate the service.");
+        }
+
+        Credentials asst_credentials = null;
+        //  Create credential and instantiate service
+        if (!string.IsNullOrEmpty(asst_username) && !string.IsNullOrEmpty(asst_password))
+        {
+            //  Authenticate using username and password
+            asst_credentials = new Credentials(asst_username, asst_password, asst_serviceUrl);
+        }
+        else if (!string.IsNullOrEmpty(asst_iamApikey))
+        {
+            //  Authenticate using iamApikey
+            TokenOptions tokenOptions2 = new TokenOptions()
```

Comment:
Did you confirm we really need to name this `tokenOptions2`? I'm thinking it should go back to `tokenOptions`.

Reply:
Nope, not needed.